lima: 1.0.7 -> 1.1.1 with extracting lima-additional-guestagents

kachick opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

- Background: https://github.com/NixOS/nixpkgs/pull/409399#discussion_r2100715385
- Closes: #409399 
- upstream tracker issue: https://github.com/lima-vm/lima/issues/3563
- Diff: https://github.com/lima-vm/lima/compare/v1.0.7...v1.1.1

Lima recommends extracting guest agents from its main package starting with version 1.1.0.

~I initially thought adding `withAdditionalGuestAgent` inputs or a `limaMinimal` package would help maintain backward compatibility in nixpkgs. However, I agree with the upstream decision to reduce the main package size. For example, the build time for the lima package on my machine decreased from 5.5 minutes to 1.5 minutes. Also, continuing to depend on deprecated make tasks makes future maintenance difficult.~

Simply extracting `lima-additional-guestagents` doesn't enable its functions in `limactl`. To address this, I've added a `withAdditionalGuestAgents` input, which creates symlinks for the new package.

Most Lima users are on macOS, but I primarily use Lima on NixOS and don't personally need these additional agents. Therefore, I'd greatly appreciate testing and feedback from macOS users and those running guest VMs with different architectures than their host.

`tree` result diff of `lima` package on x86_64-linux

```diff
--- /lima-1.0.7-binaries-task  2025-06-09 12:06:11.599635028 +0900
+++ /lima-1.1.1-native-task  2025-06-09 12:06:11.599635028 +0900
@@ -11,27 +11,45 @@
     ├── bash-completion
     │   ├── completions
     │       ├── limactl.bash
-    ├── doc
-    │   ├── lima
-    │       ├── LICENSE
-    │       ├── MAINTAINERS.md
-    │       ├── README.md
-    │       ├── ROADMAP.md
-    │       ├── SECURITY.md
-    │       ├── VERSION
-    │       ├── templates -> ../../lima/templates
     ├── fish
     │   ├── vendor_completions.d
     │       ├── limactl.fish
     ├── lima
-    │   ├── lima-guestagent.Linux-aarch64
-    │   ├── lima-guestagent.Linux-armv7l
-    │   ├── lima-guestagent.Linux-riscv64
-    │   ├── lima-guestagent.Linux-x86_64
+    │   ├── lima-guestagent.Linux-x86_64.gz
     │   ├── templates
     │       ├── README.md
+    │       ├── _default
+    │       │   ├── mounts.yaml
+    │       ├── _images
+    │       │   ├── almalinux-8.yaml
+    │       │   ├── almalinux-9.yaml
+    │       │   ├── almalinux-kitten-10.yaml
+    │       │   ├── alpine-iso.yaml
+    │       │   ├── alpine.yaml
+    │       │   ├── archlinux.yaml
+    │       │   ├── centos-stream-10.yaml
+    │       │   ├── centos-stream-9.yaml
+    │       │   ├── debian-11.yaml
+    │       │   ├── debian-12.yaml
+    │       │   ├── fedora-41.yaml
+    │       │   ├── fedora-42.yaml
+    │       │   ├── fedora.yaml
+    │       │   ├── opensuse-leap.yaml
+    │       │   ├── oraclelinux-8.yaml
+    │       │   ├── oraclelinux-9.yaml
+    │       │   ├── rocky-8.yaml
+    │       │   ├── rocky-9.yaml
+    │       │   ├── ubuntu-20.04.yaml
+    │       │   ├── ubuntu-22.04.yaml
+    │       │   ├── ubuntu-24.04.yaml
+    │       │   ├── ubuntu-24.10.yaml
+    │       │   ├── ubuntu-25.04.yaml
+    │       │   ├── ubuntu-lts.yaml
+    │       │   ├── ubuntu.yaml
     │       ├── almalinux-8.yaml
     │       ├── almalinux-9.yaml
+    │       ├── almalinux-kitten-10.yaml
+    │       ├── almalinux-kitten.yaml
     │       ├── almalinux.yaml
     │       ├── alpine-iso.yaml
     │       ├── alpine.yaml
@@ -50,6 +68,7 @@
     │       ├── docker.yaml
     │       ├── experimental
     │       │   ├── alsa.yaml
+    │       │   ├── debian-sid.yaml
     │       │   ├── gentoo.yaml
     │       │   ├── opensuse-tumbleweed.yaml
     │       │   ├── rke2.yaml
@@ -57,9 +76,12 @@
     │       │   ├── vnc.yaml
     │       │   ├── wsl2.yaml
     │       ├── faasd.yaml
+    │       ├── fedora-41.yaml
+    │       ├── fedora-42.yaml
     │       ├── fedora.yaml
     │       ├── k3s.yaml
     │       ├── k8s.yaml
+    │       ├── linuxbrew.yaml
     │       ├── opensuse-leap.yaml
     │       ├── opensuse.yaml
     │       ├── oraclelinux-8.yaml
@@ -74,10 +96,11 @@
     │       ├── ubuntu-22.04.yaml
     │       ├── ubuntu-24.04.yaml
     │       ├── ubuntu-24.10.yaml
+    │       ├── ubuntu-25.04.yaml
     │       ├── ubuntu-lts.yaml
     │       ├── ubuntu.yaml
     ├── zsh
         ├── site-functions
             ├── _limactl
 
-15 directories, 66 files
+14 directories, 90 files
```

---

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kachick committed:
lima: 1.0.7 -> 1.1.1

Diff: https://github.com/lima-vm/lima/compare/v1.0.7...v1.1.1

Since version 1.1.0, Lima separates guest agents unless they share the
same architecture.
The existing "binaries" task is retained for backward compatibility.

kachick committed:
lima: refactor and modernize

- Prefer finalAttrs
- Prefer writableTmpDirAsHomeHook
- Update meta.description
  Lima considers Linux users.
  The latest README emphasizes WSL2-like features.
- Move to by-name
- Update comment about versionCheckHook
- Simplify version check code

kachick committed:
lima: remove additional guest agents

We removed additional guest agents from the main `lima` package.
These agents were moved to the `lima-additional-guestagents` package
in the previous commit.

This change reduces the size of the `lima` package, similar to how
upstream distributes these agents. Upstream has extracted guest agents
since version v1.1.0, except for the host's architecture.

kachick committed:
lima-additional-guestagents: init at 1.1.1

kachick committed:
lima: add kachick as maintainer

kachick commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 415093`
Commit: `776295e4e942b6b4c1d1b74952836d2fc9b7ec57`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

kachick commented on pkgs/by-name/li/lima/package.nix:
```diff
@@ -0,0 +1,112 @@
+
+  buildPhase = ''
+    runHook preBuild
+    make "VERSION=v${finalAttrs.version}" "CC=${stdenv.cc.targetPrefix}cc" native
```
I tried it, but it didn't work. I understand these aren't usable with `buildGoModule`. Could you share a working example that replaces them?

kachick replied:
`rg --multiline --multiline-dotall -i 'buildGoModule.+makeFlags' pkgs/by-name`

However this results seems some matches. I'll try again

kachick replied:
A matched file is as follows:

https://github.com/NixOS/nixpkgs/blob/0a81508bd9a411b1c08eb241f990912a3ff6f874/pkgs/by-name/ru/runc/package.nix#L47-L54

I'm not sure if these wrappings are better for this short script.

kachick replied:
https://github.com/NixOS/nixpkgs/pull/415093/commits/212f9835d062d6ef22ff50bc5a655e57188c2c51

kachick commented on pkgs/by-name/li/lima/package.nix:
```diff
@@ -0,0 +1,112 @@
+  doInstallCheck = true;
+
+  # Don't use versionCheckHook for this package until Env solutions like #403971 or #411609 are available on the master branch.
+  installCheckPhase =
```
Applied in https://github.com/NixOS/nixpkgs/compare/776295e4e942b6b4c1d1b74952836d2fc9b7ec57..701f48bb80878d6f2dd71c67fe51e647eb4c6f1f

refi64 reviewed (commented):
it might be worth adding a comment into the actual .nix that mentions that you just need to install both packages (or otherwise make them part of the same symlink tree) in order for additional-guestagents to take effect? without that context, it can be a bit confusing.

kachick committed:
lima, lima-additional-guestagents: refactor to build make flags

kachick committed:
lima, lima-additional-guestagents: clarify pkg relation in longDescription

kachick committed:
lima: add withAdditionalGuestAgents and the tests

Without copying lima-additional-guestagents files, limactl does not
detect its agents even if added the package in inputs.

kachick commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 415093`
Commit: `1afe79e13ce283b5e55eef2fe0a7b0d56e5efabf`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>colima</li>
    <li>lima</li>
    <li>lima-additional-guestagents</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

kachick committed:
lima: combine execution of substituteInPlace for Makefile
